### PR TITLE
Namespace per block.

### PR DESCRIPTION
**Pull request by chriseppstein**
RFC: https://github.com/linkedin/css-blocks/issues/332

### CSS Changes

- [x] The class selector will now be defined as selecting an attribute named `class` in the namespace of the block. (conceptual change, not requiring any stylesheet parsing changes at this time).
- [x] The default namespace for the attribute selector will be dom attributes in the namespace of the block. Therefore the `state|` namespace in the state attribute selector is no longer needed.
- [x] Global states are accessed via the attribute namespace instead of a fake tag name.
- [ ] The `:scope` selector will select an element with the attribute of `scope` that is in the namespace of the block.
- [x] States named "scope" will no longer be allowed.
- [x] non-block namespaces will be handled by registering common names of `html`, `svg`, and `math` and making them forbidden names for blocks.
- [ ] Other non-block namespaces can be registered through configuration.

### Handlebars Template Syntax Changes

- [x] The namespace identifier for the default block associated with the template is `block`.
- [ ] If no element has an attribute of `block:scope` then it is automatically assigned to the template's root element.
- [x] Instead of `class="foo other-block.bar"` we would now write `block:class="foo" other-block:class="bar"`.
- [x] The `class` attribute in html namespace is forbidden. (Later we can add configuration to allow it.)

### Documentation Changes
- [ ] Document namespace concept in tutorials/readmes.
- [x] Website syntax updates.

### CODE REVIEW

**Comment by amiller-gh** [comment]
One more item to add to the TODO list: Make the word `block` a reserved word (if it's not already... I don't think it is).

Because you reference the "default" exported block as `block`, the compiler should throw if you try to `@export (whatever as block ) from "wherever.block.css"`.

**Comment by amiller-gh** [comment]
Alt: You can just change the keyword `default` to `block` for the import/export parser. 

This may be confusing for those coming from JavaScript world though – would not recommend.

File: packages/@css-blocks/core/src/BlockParser/block-intermediates.ts [hunk 1]
```diff
@@ -122,13 +114,16 @@ export function isRootNode(node: unknown): node is selectorParser.Pseudo {
 
 export const isClassNode = selectorParser.isClassName;
 
+export const RESERVED_NAMESPACES = new Set<string | undefined | true>(["html", "math", "svg"]);
+Object.freeze(RESERVED_NAMESPACES);
```
Comment (jackson-dean) [inline packages/@css-blocks/core/src/BlockParser/block-intermediates.ts]: is this left overs? it seems the freeze is called but not assigned to anything. It also doesn't seem that freeze prevents adding to a Set.

File: packages/@css-blocks/core/src/BlockSyntax/BlockSyntax.ts [hunk 2]
```diff
@@ -70,3 +67,23 @@ export const ROOT_CLASS = ":scope";
...
 // The string `default` is used throughout the system to represent the default export of a block file.
 export const DEFAULT_EXPORT = "default";
+
+// The namespace used for referencing a default block from a template
+export const DEFAULT_NAMESPACE = "block";
+
+/**
+ * Names that a block cannot have as a namespace.
+ */
+export const RESERVED_NAMESPACES = new Set<string>([DEFAULT_EXPORT, "html", "svg", "math", "mathml", "xlink", "xml", "xmlns", "xhtml"]);
```
Comment (jackson-dean) [inline packages/@css-blocks/core/src/BlockSyntax/BlockSyntax.ts]: I could just be missing something, but this looks suspiciously similar to the RESERVED_NAMESPACES const from the previous file. Are they the same or somehow different?

File: packages/@css-blocks/vscode/fixtures/app/styles/nav.block.css [hunk 3]
```diff
@@ -1,3 +1,5 @@
+@export subnav from "./subnav.block.css";
```
Comment (jackson-dean) [inline packages/@css-blocks/vscode/fixtures/app/styles/nav.block.css]: No issue, but I'm guessing these are just setup for future test? It doesn't appear these fixtures are used atm.

Comment (chriseppstein) [reply packages/@css-blocks/vscode/fixtures/app/styles/nav.block.css]: I was using them for manual testing. Adding test cases is a good idea...

File: packages/@css-blocks/core/src/BlockSyntax/BlockSyntax.ts (hunk 2 above)
Comment (chriseppstein) [reply packages/@css-blocks/core/src/BlockSyntax/BlockSyntax.ts]: Good catch. I've standardized on this one.

File: packages/@css-blocks/core/src/BlockParser/block-intermediates.ts (hunk 1 above)
Comment (chriseppstein) [reply packages/@css-blocks/core/src/BlockParser/block-intermediates.ts]: I removed this line.

**Pull request merged**